index.js

```diff
@@ -70,63 +72,69 @@ Client.prototype._ref = function () {
   })
 }
 
-Client.prototype._write = function (obj, cb) {
+Client.prototype._write = function (obj, enc, cb) {
   if (this._destroyed) {
-    this.emit('error', new Error('data sent to destroyed Elastic APM client'))
-    return
+    this.emit('error', new Error('write called on destroyed Elastic APM client'))
+    cb()
+  } else if (obj === flush) {
+    if (this._active) {
+      this._onflushed = cb
+      this._chopper.chop()
+    } else {
+      this._chopper.chop(cb)
+    }
+  } else {
+    this._stream.write(obj, cb)
   }
-  return this._stream.write(obj, cb)
 }
 
 Client.prototype.sendSpan = function (span, cb) {
-  return this._write({span}, cb)
+  return this.write({span}, cb)
 }
 
 Client.prototype.sendTransaction = function (transaction, cb) {
-  return this._write({transaction}, cb)
+  return this.write({transaction}, cb)
 }
 
 Client.prototype.sendError = function (error, cb) {
-  return this._write({error}, cb)
+  return this.write({error}, cb)
 }
 
 Client.prototype.flush = function (cb) {
   if (this._destroyed) {
     this.emit('error', new Error('flush called on destroyed Elastic APM client'))
+    if (cb) process.nextTick(cb)
     return
   }
-  if (this._ended) return // TODO: call bacllback?
 
-  if (cb && this._active) {
-    // if there's an outgoing http request, queue the callback to be called
-    // when we're sure that the data has left the system
-    this._onflushed = cb
-    this._chopper.chop()
-  } else {
-    this._chopper.chop(cb)
-  }
```

Comment:
What's the reason for moving this into the `_write` logic?

Reply:
We write the special "flush" signal so that the order of writes and flushes are kept. If we where to just flush the client right here, the internal Writable buffer might still contain data that hasn't yet been given to the _write function.

I've added a comment to explain this.